### PR TITLE
ao_opensles: rework the heuristic of buffer/enqueue size setting

### PR DESCRIPTION
**Pull request by tomty89**

### CODE REVIEW

**Comment by tomty89** [comment]
@jeeb @sfan5 any comments/doubts? would be great if this can be merged before 0.29.

File: audio/out/ao_opensles.c [hunk 1]
```diff
@@ -155,17 +153,39 @@ static int init(struct ao *ao)
...
 
-    if (p->cfg_frames_per_buffer)
-        ao->device_buffer = p->cfg_frames_per_buffer;
-    else
-        ao->device_buffer = ao->samplerate * DEFAULT_BUFFER_SIZE_MS / 1000;
-    p->buffer_size = ao->device_buffer * ao->channels.num *
+    if (p->buffer_size_in_ms) {
+        ao->device_buffer = ao->samplerate * p->buffer_size_in_ms / 1000;
+        /*
+          As the purpose of buffer_size_in_ms is to request a specific
```
Comment (sfan5) [inline audio/out/ao_opensles.c]: indent should be 4 spaces here (?)

**Comment by sfan5** [comment]
Can't comment on the code much, but it certainly works fine on my device.
Does this make the "ideal" [sample rate / buffer selection](https://github.com/mpv-android/mpv-android/blob/master/app/src/main/java/is/xyz/mpv/MPVView.kt#L56) in mpv-android obsolete?

**Comment by tomty89** [comment]
Well that's not one of the intentions of this PR. But the fact it allows users to reduce the soft buffer size it requests, it serves that purpose in some sense (in a more proper/useful way).

But the PR is created with the consideration that you are not going to drop the selection. The only behavioral change in mpv-android is, the default soft buffer will be 250ms (which was chosen with reasons/data) instead of 200ms (the arbitrary generic default in mpv).

That said, I would be more than happy if you decide to drop the "buffer selection", as it was never a right thing to do (for various reasons; which I am not gonna list here as I don't want to divert the discussion; I can open a PR for that if you like and we can discuss there).

It might be worth mentioning that I also created a PR that introduces a new AO which makes use of the AAudio API (#5945), which finally allows the buffer thing to be done dynamically in a sensical way.

**Comment by sfan5** [comment]
So, we'd drop `opensles-frames-per-buffer` but still set `audio-samplerate`? Is that what you had in mind?

**Comment by tomty89** [comment]
@sfan5 Sample rate setting is yet another story. That one is more of a matter of taste.

The resampler in ffmpeg may be more preferable, and sample rate that matches with the audio sink is indeed the criterion of making the track a "fast track" (whether that matters is in doubt though).

However the problem is that, it also carries the assumption that we do not change audio device during playback. If we do, not only that the rate may not be optimal anymore but double resampling could also occur in that case.

mpv-android is using the wrong method to get the rate anyway, as the AudioManager method provides static/garbage value. AudioTrack method should be used instead.

Personally I don't think we should do sample rate selection. The bottom line is that it should be optional.

**Comment by tomty89** [comment]
Btw I think we should keep the `--opensles-frames-per-enqueue` option, even if mpv-android is not going to use it anymore.

For one it might be useful for testing/debugging in the future. Also I think it's a good to show the relation between Enqueue size and soft buffer size.

In case at some point we decided that firing multiple Enqueue() in resume() should be made possible, the current code in this PR would make things much easier and clearer (as the buffer size should be equal to the Enqueue size multiply by number of Enqueue() we fire).

**Pull request merged**